config/config.go

```diff
@@ -0,0 +1,459 @@
+	nameParts := make([][]string, len(fieldParts))
+	for index, field := range fieldParts {
+		parts := []string{}
+		lastIsCapital := true
+		lastIndex := 0
+		for i, char := range field {
+			currentIsCapital := char >= 'A' && char <= 'Z'
+			if i > 0 && lastIsCapital != currentIsCapital {
+				end := 0
+				if currentIsCapital {
+					end = i
+				} else {
+					end = i - 1
+				}
+				if end > lastIndex {
+					parts = append(parts, field[lastIndex:end])
+					lastIndex = end
+				}
+			}
+			lastIsCapital = currentIsCapital
+		}
+		if lastIndex < len(field) {
+			parts = append(parts, field[lastIndex:])
+		}
+		nameParts[index] = parts
+	}
+	cf := &configField{
+		pointer:     pointer,
+		shortFlag:   shortFlag,
+		description: desc,
+	}
+	for i, parts := range nameParts {
+		if i > 0 {
+			cf.key += "."
+		}
+		for j, part := range parts {
+			if j == 0 {
+				part = strings.ToLower(part)
+			}
+			cf.key += part
+		}
+	}
+	for i, parts := range nameParts {
+		if i > 0 {
+			cf.longFlag += "-"
+		}
+		cf.longFlag += strings.Join(parts, "-")
+	}
+	cf.longFlag = strings.ToLower(cf.longFlag)
+	for i, parts := range nameParts {
+		if i > 0 {
+			cf.env += "_"
+		}
+		cf.env += strings.Join(parts, "_")
+	}
+	cf.env = strings.ToUpper(cf.env)
+	if _, ok := s.fields[cf.key]; ok {
+		panic(errors.InternalServerError.Error("${key} has been registered", cf.key))
+	}
+	s.fields[cf.key] = cf
+	return s
+}
+
+// Execute runs nirvana server.
+func (s *starter) Execute(descriptors ...definition.Descriptor) error {
+	cfg := nirvana.NewDefaultConfig("", 80)
+	cfg.Configure(nirvana.Descriptor(descriptors...))
+	return s.Command(cfg).Execute()
+}
+
+// ExecuteWithConfig runs nirvana server from a custom config.
+func (s *starter) ExecuteWithConfig(cfg *nirvana.Config) error {
+	return s.Command(cfg).Execute()
+}
+
+// Command returns a command for starter.
+func (s *starter) Command(cfg *nirvana.Config) *cobra.Command {
+	cmd := &cobra.Command{
+		Use: s.name,
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg.Logger.Info(banner)
+			// Restore configs.
+			for _, f := range s.fields {
+				val := reflect.ValueOf(f.pointer).Elem().Interface()
+				Set(f.key, val)
+			}
+			for _, plugin := range s.plugins {
+				if err := plugin.Configure(cfg); err != nil {
+					cfg.Logger.Fatalf("Failed to install plugin %s: %s", plugin.Name(), err.Error())
+				}
+			}
+			cfg.Logger.Infof("Listening on %s:%d", cfg.IP, cfg.Port)
+			if err := nirvana.NewServer(cfg).Serve(); err != nil {
+				cfg.Logger.Fatal(err)
+			}
+		},
+	}
+	fs := cmd.Flags()
+	for _, f := range s.fields {
+		var value, envValue interface{} = nil, nil
+		env := os.Getenv(f.env)
+		switch v := f.pointer.(type) {
+		case *uint8:
+			if IsSet(f.key) {
+				value = GetUint8(f.key)
+			}
+			if env != "" {
+				envValue = cast.ToUint8(env)
+			}
+			desc, val := chooseValue(f.key, value, f.env, envValue, *v)
+			fs.Uint8VarP(v, f.longFlag, f.shortFlag, val.(uint8), f.description+desc)
```

Comment:
can you explain the priorities of different config source: 
- explicit calls to set
- flag
- env
- default value
- config file

Reply:
Flag > ENV > Config file > Default